Index: mingw-ocaml/build/findlib/src/findlib/Makefile
===================================================================
--- mingw-ocaml.orig/build/findlib/src/findlib/Makefile	2013-10-13 08:35:04.805672918 -0300
+++ mingw-ocaml/build/findlib/src/findlib/Makefile	2013-10-13 16:21:39.457686493 -0300
@@ -11,8 +11,8 @@
 
 NAME = findlib
 
-OCAMLC = ocamlc
-OCAMLOPT = ocamlopt
+OCAMLC = ocamlc -I +compiler-libs
+OCAMLOPT = ocamlopt -I +compiler-libs
 OCAMLDEP = ocamldep
 OCAMLLEX = ocamllex
 CAMLP4O =  camlp4 pa_o.cmo pa_op.cmo pr_o.cmo --
