Index: mingw-ocaml/build/ocaml/byterun/win32.c
===================================================================
--- mingw-ocaml.orig/build/ocaml/byterun/win32.c	2013-10-13 08:55:19.000000000 -0300
+++ mingw-ocaml/build/ocaml/byterun/win32.c	2013-10-13 08:58:24.000000000 -0300
@@ -55,6 +55,26 @@
 };
 #endif
 
+/* XXX including <io.h> gets ../byterun/io.h for some reason.
+ * Including the real io.h using the full path fails because of
+ * some strange bug in the system header file itself.  Give up and
+ * just define _finddata_t explicitly here.
+ */
+#ifndef _FSIZE_T_DEFINED
+typedef unsigned long   _fsize_t;
+#define _FSIZE_T_DEFINED
+
+struct _finddata_t
+{
+  unsigned        attrib;
+  time_t          time_create;
+  time_t          time_access;
+  time_t          time_write;
+  _fsize_t        size;
+  char            name[FILENAME_MAX];
+};
+#endif
+
 #ifndef S_ISREG
 #define S_ISREG(mode) (((mode) & S_IFMT) == S_IFREG)
 #endif
