Index: mingw-ocaml/build/ocaml/asmrun/i386.S
===================================================================
--- mingw-ocaml.orig/build/ocaml/asmrun/i386.S	2013-03-22 16:08:07.554296392 -0500
+++ mingw-ocaml/build/ocaml/asmrun/i386.S	2013-03-22 17:20:07.749047630 -0500
@@ -76,5 +76,8 @@
         call mcount; \
         popl %ebp; CFI_ADJUST(-4)
+#elif defined(SYS_mingw)
+#define PROFILE_CAML
+#define PROFILE_C
 #elif defined(SYS_bsd_elf)
 #define PROFILE_CAML \
         pushl %ebp; CFI_ADJUST(4); \
